fabtests/prov/efa/src/rdm_ep_rnr_retry.c

```diff
@@ -0,0 +1,215 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * This program tests the RNR retry counter reset via fi_setopt.
+ * When running the test, use `-R` option to specify RNR retry counter.
+ * The valid values are 0 - 7 (7 inidicates infinite retry on firmware).
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <shared.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_ext.h>
+
+static int poll_rnr_cq_error(void)
+{
+	struct fi_cq_data_entry comp;
+	struct fi_cq_err_entry comp_err;
+	int total_send, expected_rnr_error;
+	int ret, i, cnt, rnr_flag;
+
+	rnr_flag = 0;
+	expected_rnr_error = 1;
+	cnt = total_send = fi->rx_attr->size + expected_rnr_error;
+
+	for (i = 0; i < total_send; i++) {
+		do {
+			ret = fi_send(ep, tx_buf, 32, mr_desc, remote_fi_addr, &tx_ctx);
+			if (ret < 0 && ret != -FI_EAGAIN) {
+				FT_PRINTERR("fi_send", -ret);
+				return ret;
+			}
+		} while (ret == -FI_EAGAIN);
+	}
+
+	do {
+		ret = fi_cq_read(txcq, &comp, 1);
+		if (ret == 1) {
+			cnt--;
+		} else if (ret == -FI_EAVAIL) {
+			ret = fi_cq_readerr(txcq, &comp_err, FI_SEND);
+			if (ret < 0 && ret != -FI_EAGAIN) {
+				FT_PRINTERR("fi_cq_readerr", -ret);
+				return ret;
+			} else if (ret == 1 && comp_err.err == FI_ENORX) {
+				cnt--;
+				rnr_flag = 1;
+				printf("Got error CQ entry as expected: %d, %s\n",
+					comp_err.err, fi_strerror(comp_err.err));
+			}
+		} else if (ret < 0 && ret != -FI_EAGAIN) {
+			FT_PRINTERR("fi_cq_read", -ret);
+			return ret;
+		}
+	} while (cnt);
+
+	return (rnr_flag) ? 0 : -FI_EINVAL;
+}
+
+static int run(size_t rnr_retry)
+{
+	int ret;
+
+	ret = ft_init();
+	if (ret) {
+		FT_PRINTERR("ft_init", -ret);
+		return ret;
+	}
+
+	ret = ft_init_oob();
+	if (ret) {
+		FT_PRINTERR("ft_init_oob", -ret);
+		return ret;
+	}
+
+	ret = ft_getinfo(hints, &fi);
+	if (ret) {
+		FT_PRINTERR("ft_getinfo", -ret);
+		return ret;
+	}
+
+	ret = ft_open_fabric_res();
+	if (ret) {
+		FT_PRINTERR("ft_open_fabric_res", -ret);
+		return ret;
+	}
+
+	ret = ft_alloc_active_res(fi);
+	if (ret) {
+		FT_PRINTERR("ft_alloc_active_res", -ret);
+		return ret;
+	}
+
+	fprintf(stdout, "Setting RNR retry count to %zu ...\n", rnr_retry);
+	ret = fi_setopt(&ep->fid, FI_OPT_ENDPOINT, FI_OPT_EFA_RNR_RETRY, &rnr_retry, sizeof(rnr_retry));
+	if (ret) {
+		FT_PRINTERR("fi_setopt", -ret);
+		return ret;
+	}
+	fprintf(stdout, "RNR retry count has been set to %zu.\n", rnr_retry);
+
+	ret = ft_enable_ep(ep, eq, av, txcq, rxcq, txcntr, rxcntr);
+	if (ret) {
+		FT_PRINTERR("ft_enable_ep_recv", -ret);
+		return ret;
+	}
+
+	ret = ft_init_av();
+	if (ret) {
+		FT_PRINTERR("ft_init_av", -ret);
+		return ret;
+	}
+	/* client does fi_send and then poll CQ to get error (FI_ENORX) CQ entry */
+	if (opts.dst_addr) {
+		ret = poll_rnr_cq_error();
+		if (ret) {
+			FT_PRINTERR("pingpong", -ret);
+			return ret;
+		}
+	}
+
+	/* use ft_reset_oob as a sync mechanism, so we know client finishes poll_rnr_cq_error */
```

Comment:
In the comment, can you explain what is special about `ft_reset_oob` and why it is used?